cpp/oneapi/dal/algo/kmeans/test/spmd_backend_fixture.hpp

```diff
@@ -0,0 +1,669 @@
+                                      std::int64_t cluster_count,
+                                      std::int64_t max_iteration_count,
+                                      float_t accuracy_threshold,
+                                      float_t ref_objective_function = -1.0,
+                                      bool test_convergence = false) {
+        CAPTURE(cluster_count);
+
+        INFO("create descriptor")
+        const auto kmeans_desc =
+            get_descriptor(cluster_count, max_iteration_count, accuracy_threshold);
+
+        INFO("run training");
+        const auto train_result = this->train(kmeans_desc, data, initial_centroids);
+        const auto model = train_result.get_model();
+
+        auto match_map = array<float_t>::zeros(cluster_count);
+        find_match_centroids(ref_centroids,
+                             model.get_centroids(),
+                             ref_centroids.get_column_count(),
+                             match_map);
+        check_train_result(kmeans_desc,
+                           train_result,
+                           match_map,
+                           ref_centroids,
+                           ref_responses,
+                           test_convergence);
+        INFO("run inference");
+        const auto infer_result = this->infer(kmeans_desc, model, data);
+        check_infer_result(kmeans_desc,
+                           infer_result,
+                           match_map,
+                           ref_responses,
+                           ref_objective_function);
+    }
+
+    void check_on_gold_data() {
+        const auto table_id = this->get_homogen_table_id();
+        const auto data = gold_dataset::get_data().get_table(table_id);
+        const auto initial_centroids = gold_dataset::get_initial_centroids().get_table(table_id);
+        const auto expected_centroids = gold_dataset::get_expected_centroids().get_table(table_id);
+        const auto expected_responses = gold_dataset::get_expected_responses().get_table(table_id);
+
+        const std::int64_t cluster_count = gold_dataset::get_cluster_count();
+        const std::int64_t max_iteration_count = 100;
+        const float_t accuracy_threshold = 0.0;
+
+        INFO("create descriptor")
+        const auto kmeans_desc =
+            get_descriptor(cluster_count, max_iteration_count, accuracy_threshold);
+
+        INFO("run training");
+        const auto train_result = this->train(kmeans_desc, data, initial_centroids);
+        const auto model = train_result.get_model();
+        const auto centroids = model.get_centroids();
+
+        auto match_map = array<float_t>::zeros(cluster_count);
+        find_match_centroids(expected_centroids,
+                             centroids,
+                             expected_centroids.get_column_count(),
+                             match_map);
+
+        INFO("check if centroids close to gold") {
+            const double rel_tol = 1e-7;
```

Comment:
Do we need different rel_tol values for double and for float?
The same question for obj_ref_tol, dbi_ref_tol values further in this file.

Reply:
It seems to me, there is no need to use different values. The same values are used for CPU/GPU/SPMD. Floating point values aren't the main source of discrepancy.